src/textual/widgets/text_input.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+
+from rich.console import RenderableType
+from rich.text import Text
+
+from textual import events
+from textual._types import MessageTarget
+from textual.geometry import Size
+from textual.keys import Keys
+from textual.message import Message
+from textual.reactive import Reactive
+from textual.widget import Widget
+
+
+class TextInputBase(Widget):
+    ALLOW_PROPAGATE = {"tab", "shift+tab"}
+
+    current_text = Reactive("", layout=True)
+    cursor_index = Reactive(0)
+
+    class Changed(Message, bubble=True):
```

Comment:
I'm thinking of potentially adding a `validation_result` field or something like that to this message, so that for each change, the message contains the value and some object which describes that value's validity.